contracts/token/ERC721/ERC721Token.sol

```diff
@@ -52,147 +51,91 @@ contract ERC721Token is ERC721 {
-    if (approvedFor(_tokenId) != 0) {
-      clearApproval(msg.sender, _tokenId);
-    }
-    removeToken(msg.sender, _tokenId);
-    Transfer(msg.sender, 0x0, _tokenId);
-  }
-
-  /**
-   * @dev Tells whether the msg.sender is approved for the given token ID or not
-   * This function is not private so it can be extended in further implementations like the operatable ERC721
-   * @param _owner address of the owner to query the approval of
-   * @param _tokenId uint256 ID of the token to query the approval of
-   * @return bool whether the msg.sender is approved for the given token ID or not
-   */
-  function isApprovedFor(address _owner, uint256 _tokenId) internal view returns (bool) {
-    return approvedFor(_tokenId) == _owner;
+  function tokenOfOwnerByIndex(address _owner, uint256 _index) public view returns (uint256) {
+    require(_index < balanceOf(_owner));
+    return ownedTokens[_owner][_index];
   }
 
   /**
-  * @dev Internal function to clear current approval and transfer the ownership of a given token ID
-  * @param _from address which you want to send tokens from
-  * @param _to address which you want to transfer the token to
-  * @param _tokenId uint256 ID of the token to be transferred
+  * @dev Gets the total amount of tokens stored by the contract
+  * @return uint256 representing the total amount of tokens
   */
-  function clearApprovalAndTransfer(address _from, address _to, uint256 _tokenId) internal {
-    require(_to != address(0));
-    require(_to != ownerOf(_tokenId));
-    require(ownerOf(_tokenId) == _from);
-
-    clearApproval(_from, _tokenId);
-    removeToken(_from, _tokenId);
-    addToken(_to, _tokenId);
-    Transfer(_from, _to, _tokenId);
+  function totalSupply() public view returns (uint256) {
+    return allTokens.length;
   }
 
   /**
-  * @dev Internal function to clear current approval of a given token ID
-  * @param _tokenId uint256 ID of the token to be transferred
+  * @dev Gets the token ID at a given index of all the tokens in this contract
+  * @dev Reverts if the index is greater or equal to the total number of tokens
+  * @param _index uint256 representing the index to be accessed of the tokens list
+  * @return uint256 token ID at the given index of the tokens list
   */
-  function clearApproval(address _owner, uint256 _tokenId) private {
-    require(ownerOf(_tokenId) == _owner);
-    tokenApprovals[_tokenId] = 0;
-    Approval(_owner, 0, _tokenId);
+  function tokenByIndex(uint256 _index) public view returns (uint256) {
+    require(_index < totalSupply());
+    return allTokens[_index];
   }
 
   /**
   * @dev Internal function to add a token ID to the list of a given address
   * @param _to address representing the new owner of the given token ID
   * @param _tokenId uint256 ID of the token to be added to the tokens list of the given address
   */
-  function addToken(address _to, uint256 _tokenId) private {
+  function addToken(address _to, uint256 _tokenId) internal {
```

Comment:
This should be implemented calling `super.addToken`. Although I understand there will be redundancy between `ownedTokens` and `ownedTokensCount`, we can't leave that state variable unused because some other contract in the hierarchy (even `ERC721BasicToken`) may be using it.

Is there some other reason I'm not seeing?

Reply:
Good point, I'm using super to keep the state var in sync, and removing the overload.